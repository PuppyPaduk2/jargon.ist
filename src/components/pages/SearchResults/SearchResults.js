// @flow

import React, { Fragment, PureComponent } from 'react';
import Link, { navigateTo } from 'gatsby-link';
import { connectStateResults } from 'react-instantsearch/connectors';

import { Container, Section } from '../../ui';

import { type Jargon } from '../../../types/jargon';
import { type Tag } from '../../../types/tag';

type Props = {
  allSearchResults?: {
    jargons?: {
      query: string,
      hits: Array<Jargon>,
    },
    tags?: {
      query: string,
      hits: Array<Tag>,
    },
  },
};

export class SearchResultsWrapper extends PureComponent<Props> {
  componentDidMount() {
    this.navigateIfNecessary();
  }

  componentDidUpdate() {
    this.navigateIfNecessary();
  }

  navigateIfNecessary = () => {
    if (!this.props.allSearchResults) {
      return;
    }

    const { jargons, tags } = this.props.allSearchResults;

    if (!jargons || !tags) {
      return;
    }

    if (
      jargons.hits.length === 1 &&
      tags.hits.length === 0 &&
      jargons.query.toLocaleLowerCase() === jargons.hits[0].title.toLocaleLowerCase()
    ) {
      navigateTo(`/${jargons.hits[0].slug}`);
    }

    if (
      jargons.hits.length === 0 &&
      tags.hits.length === 1 &&
      tags.query.toLocaleLowerCase() === tags.hits[0].title.toLocaleLowerCase()
    ) {
      navigateTo(`/k/${tags.hits[0].slug}`);
    }
  };

  render() {
    const { jargons, tags } = this.props.allSearchResults || {};

    return (
      <Fragment>
        <Section className="u-bg-steel-gray">
          <Container>
            <h1 className="u-color-primary u-clear-gap">Arama Sonuçları</h1>
          </Container>
        </Section>
        <Section>
          <Container>
            <div className="row">
              <div className="col col--lg-6 u-gap-bottom-xlarge@md-down">
                <h2 className="u-color-primary u-clear-gap-top">Jargonlar</h2>
                {jargons &&
                  jargons.hits.length > 0 &&
                  jargons.hits.map(jargon => (
                    <div key={jargon.slug} className="u-gap-bottom-2xsmall">
                      <Link to={`/${jargon.slug}`} className="c-link-secondary">
                        {jargon.title}
                      </Link>
                    </div>
                  ))}
                {jargons && jargons.hits.length === 0 && <p>Herhangi bir jargon bulunamadı.</p>}
              </div>

              <div className="col col--lg-6">
                <h2 className="u-color-primary u-clear-gap-top">Konular</h2>
                {tags &&
                  tags.hits.length > 0 &&
                  tags.hits.map(tag => (
                    <div key={tag.slug} className="u-gap-bottom-2xsmall">
                      <Link to={`/k/${tag.slug}`} className="c-link-secondary">
                        {tag.title}
                      </Link>
                    </div>
                  ))}
                {tags && tags.hits.length === 0 && <p>Herhangi bir konu bulunamadı.</p>}
              </div>
            </div>
          </Container>
        </Section>
      </Fragment>
    );
  }
}

export default connectStateResults(SearchResultsWrapper);
