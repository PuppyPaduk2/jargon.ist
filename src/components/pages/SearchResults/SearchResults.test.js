// @flow

import React from 'react';
import { shallow, mount } from 'enzyme';
import { navigateTo } from 'gatsby-link';

import jargonsFixture from '../../../../__fixtures__/jargons';
import tagsFixture from '../../../../__fixtures__/tags';

import { SearchResultsWrapper } from './SearchResults';

jest.mock('gatsby-link');

beforeEach(() => {
  navigateTo.mockClear();
});

it('renders correctly', () => {
  const allSearchResults = {
    jargons: {
      query: 'p',
      hits: jargonsFixture,
    },
    tags: {
      query: 'p',
      hits: tagsFixture,
    },
  };

  const wrapper = shallow(<SearchResultsWrapper allSearchResults={allSearchResults} />);

  expect(wrapper).toMatchSnapshot();
});

it('navigates to related jargon if it is the only result', () => {
  const allSearchResults = {
    jargons: {
      query: 'Example Jargon 1',
      hits: [
        {
          html: '<p>Hello world</p>',
          title: 'Example Jargon 1',
          slug: 'example-jargon-1',
          tags: [{ title: 'tag1', slug: 'tag1' }, { title: 'tag2', slug: 'tag2' }],
          sameWith: ['sameWith1', 'sameWith2'],
        },
      ],
    },
    tags: {
      query: 'Example Jargon 1',
      hits: [],
    },
  };

  mount(<SearchResultsWrapper allSearchResults={allSearchResults} />);

  expect(navigateTo).toHaveBeenCalledWith('/example-jargon-1');
});

it('navigates to related tag if it is the only result', () => {
  const allSearchResults = {
    jargons: {
      query: 'Tag 1',
      hits: [],
    },
    tags: {
      query: 'Tag 1',
      hits: [
        {
          title: 'tag 1',
          slug: 'tag-1',
        },
      ],
    },
  };

  mount(<SearchResultsWrapper allSearchResults={allSearchResults} />);

  expect(navigateTo).toHaveBeenCalledWith('/k/tag-1');
});
