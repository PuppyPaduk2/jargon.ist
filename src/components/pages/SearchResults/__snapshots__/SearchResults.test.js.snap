// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<React.Fragment>
  <Section
    className="u-bg-steel-gray"
  >
    <Container
      className=""
    >
      <h1
        className="u-color-primary u-clear-gap"
      >
        Arama Sonuçları
      </h1>
    </Container>
  </Section>
  <Section
    className=""
  >
    <Container
      className=""
    >
      <div
        className="row"
      >
        <div
          className="col col--lg-6 u-gap-bottom-xlarge@md-down"
        >
          <h2
            className="u-color-primary u-clear-gap-top"
          >
            Jargonlar
          </h2>
          <div
            className="u-gap-bottom-2xsmall"
            key="example-jargon-1"
          >
            <GatsbyLink
              className="c-link-secondary"
              to="/example-jargon-1"
            >
              Example Jargon 1
            </GatsbyLink>
          </div>
          <div
            className="u-gap-bottom-2xsmall"
            key="example-jargon-2"
          >
            <GatsbyLink
              className="c-link-secondary"
              to="/example-jargon-2"
            >
              Example Jargon 2
            </GatsbyLink>
          </div>
        </div>
        <div
          className="col col--lg-6"
        >
          <h2
            className="u-color-primary u-clear-gap-top"
          >
            Konular
          </h2>
          <div
            className="u-gap-bottom-2xsmall"
            key="tag1"
          >
            <GatsbyLink
              className="c-link-secondary"
              to="/k/tag1"
            >
              tag1
            </GatsbyLink>
          </div>
          <div
            className="u-gap-bottom-2xsmall"
            key="tag2"
          >
            <GatsbyLink
              className="c-link-secondary"
              to="/k/tag2"
            >
              tag2
            </GatsbyLink>
          </div>
        </div>
      </div>
    </Container>
  </Section>
</React.Fragment>
`;
